# ESG financial Insights Generator 

The system will retrieve the latest ESG reports and regulatory documents to generate concicse, insightful summaries answering ESG-related queries for a specific ompany. 

Problem statement: How can a tool summarise ESG related news of each company? 

Background: 
This is a continuation from assignment1.Based on these newly sourced company ESG reports, and the the previous classification method from assignment1, I can classify these chunks of text into the E,S and G categories, so that model is able to better generate a response related to one of the particular categories. For example, the query could be "What are the recent carbon emission goals of Pfizer?" Since in the data loading portion of this assignmnet I have added Pfizer's ESG report that details ESG strategies, and "carbon emission" is under the "environmental" category, based on these contexts, the model is able to get the specific document chunk context and metadata (the categories) to answer this question.

In [2]:
import os
import ocrmypdf
from PyPDF2 import PdfReader
import re
import json
import fitz  # PyMuPDF for PDF extraction
import chromadb  # Vector Database
from tqdm import tqdm
import logging
import requests
import time
import pandas as pd
import numpy as np
import torch 
from torch import nn
from torch.optim import AdamW  
from torch.utils.data import Dataset, DataLoader
import numpy as np
import pandas as pd 
from io import BytesIO
from transformers import AutoModelForSequenceClassification, AutoModelForCausalLM, AutoTokenizer, pipeline, BertTokenizer, BertModel, Trainer, TrainingArguments
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import spacy
from spacy import displacy
### if u got strong gpu w cuda i recommend changing to gpu, average laptop cpu takes too long *cough* mac book users
torch.set_default_device("cpu")
if torch.cuda.is_available():
    torch.set_default_device("cuda")
    print("running on cuda")
import random
import json
import google.generativeai as genai
google_api_key = "AIzaSyCutzQsZEOJUQgHwcvjtPNiLFbgyxOfmko"
from openai import OpenAI
API_KEY = "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" 

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from googlesearch import search
from fuzzywuzzy import fuzz 

# Data Sourcing & Preprocessing
- ESG news articles (This was pulled from the previous project, using APIs and webscraping, so I transferred the downloaded files over.)
- Company ESG pdf reports

## Automate pulling pdfs online

In [14]:
genai.configure(api_key=google_api_key)
model = genai.GenerativeModel('gemini-2.0-flash')
gemini_eval = model.generate_content(
    contents= "Can you randomly generate a list of company names that come from different sizes, based in different countries, and of various industries. For example, companies that could be generated are [Pfizer, Apple, Uber, LVMH], etc where they are in the healthcare, technology, transport, FMCG and based in different countries. In your response, number the list like so 1.Pfizer, 2.Apple "
)
print(gemini_eval.text)
company_names = [name for name in re.findall(r"\*\*(.+?)\*\*", gemini_eval.text)  if "Disclaimer" not in name]
company_list = company_names[:3]
years=[2024, 2023]
print(company_list)

Okay, here's a randomly generated list of company names with varying sizes, industries, and countries:

1.  **Samsung** (South Korea, Technology, Large)
2.  **Nestlé** (Switzerland, Food & Beverage, Large)
3.  **Tata Consultancy Services (TCS)** (India, IT Services, Large)
4.  **Adidas** (Germany, Apparel & Sportswear, Large)
5.  **Toyota** (Japan, Automotive, Large)
6.  **Shopify** (Canada, E-commerce, Large)
7.  **Eni** (Italy, Energy, Large)
8.  **Banco Santander** (Spain, Banking, Large)
9.  **Woolworths Group** (Australia, Retail, Large)
10. **Unilever** (United Kingdom/Netherlands, Consumer Goods, Large)
11. **Novo Nordisk** (Denmark, Pharmaceutical, Large)
12. **Siemens** (Germany, Manufacturing, Large)
13. **BP** (United Kingdom, Energy, Large)
14. **TikTok (ByteDance)** (China, Social Media, Large)
15. **Hermès** (France, Luxury Goods, Large)

['Samsung', 'Nestlé', 'Tata Consultancy Services (TCS)']


In [15]:
pdf_links = []
for company in company_list:
    for year in years:
        query = f"{company} {year} ESG report filetype:pdf"
        
        for url in search(query, num_results=1):
            if url.endswith(".pdf"):  
                pdf_links.append(url)
print(pdf_links)

['https://www.samsung.com/global/sustainability/media/pdf/Samsung_Electronics_Sustainability_Report_2024_ENG.pdf', 'https://www.samsung.com/global/sustainability/media/pdf/Samsung_Electronics_Sustainability_Report_2023_ENG.pdf', 'https://www.nestle.com/sites/default/files/2025-02/non-financial-statement-2024.pdf', 'https://www.nestle.com/sites/default/files/2024-02/creating-shared-value-sustainability-report-2023-en.pdf', 'https://www.tcs.com/content/dam/tcs/pdf/discover-tcs/corporate-sustainability/tcs-carbon-reduction-plan-aug-2024.pdf', 'https://www.tcs.com/content/dam/global-tcs/en/pdfs/investors/esg/2023-24/fy24-qes-mea.pdf']


### pdf image to text

In [16]:
def extract_company_and_year(pdf_url):
    """
    Extract the company name and year from the URL or filename.
    Assuming the URL contains the company name and year in a predictable format like: 
    Assumes company name is right after 'https://' and the year is just before '.pdf'.
    """
    match = re.search(r'https://(?:www\.)?([a-zA-Z0-9-]+).*?(\d{4}(?:-\d{4})?)\.pdf', pdf_url)
    if match:
        company_name = match.group(1)  # The company name
        year = match.group(2)  # The year
        return company_name, year
    else:
        print(f"Failed to name and year extract from URL: {pdf_url}")
        return None, None

def process_pdf_from_url(pdf_url, output_pdf_path):
    """
    Processes a PDF from a URL, applies OCR, and saves the OCR version.
    """
    if os.path.exists(output_pdf_path):
        print(f"Skipping '{pdf_url}': already processed.")
        return

    print(f"Processing '{pdf_url}'...")
    try:
        response = requests.get(pdf_url)
        response.raise_for_status()  # Check for HTTP errors
        pdf_data = BytesIO(response.content)  # Convert the PDF to BytesIO stream

        # Apply OCR to the PDF data
        ocrmypdf.ocr(pdf_data, output_pdf_path, force_ocr=True)
        print(f"Saved OCR version to '{output_pdf_path}'.")
    except Exception as e:
        print(f"Error processing '{pdf_url}': {e}")

def process_pdfs_from_urls(pdf_urls, output_root):
    for pdf_url in pdf_urls:
        company_name, year = extract_company_and_year(pdf_url)
        print(f"Processing PDF from compay: {company_name}, with URL: {pdf_url}")
        output_pdf_path = os.path.join(output_root, f"{pdf_url.split('/')[-1].split('.')[0]}_ocr.pdf")
        process_pdf_from_url(pdf_url, output_pdf_path)

In [19]:
output_root = r"datasets"
if not os.path.exists(output_root):
    os.makedirs(output_root)
process_pdfs_from_urls(pdf_links, output_root)

Failed to name and year extract from URL: https://www.samsung.com/global/sustainability/media/pdf/Samsung_Electronics_Sustainability_Report_2024_ENG.pdf
Processing PDF from compay: None, with URL: https://www.samsung.com/global/sustainability/media/pdf/Samsung_Electronics_Sustainability_Report_2024_ENG.pdf
Processing 'https://www.samsung.com/global/sustainability/media/pdf/Samsung_Electronics_Sustainability_Report_2024_ENG.pdf'...


Output()

This PDF has a fillable form. Chances are it is a pure digital document that does not need OCR.


Output()

[tesseract] lots of diacritics - possibly poor OCR

[tesseract] lots of diacritics - possibly poor OCR

page has no images - all vector content will be rasterized at 400 DPI, losing some resolution and likely increasing
file size. Use --oversample to adjust the DPI.

Some input metadata could not be copied because it is not permitted in PDF/A. You may wish to examine the output PDF's XMP metadata.


KeyboardInterrupt: 

### pdf to text reader

In [20]:
def convert_ocr_pdf_to_text(input_pdf_path):
    """
    Extract text from an OCR-based PDF, clean it, and save as a JSON list of sentences.
    """
    def clean_page_text(text):
        """
        Remove headers, footers, and page numbers from extracted text.
        """
        lines = text.split("\n")
        if len(lines) > 2:
            lines = lines[2:]
        lines = [line for line in lines if not re.match(r'^(Page\s*\d+|\d+|P\.\s*\d+)$', line.strip(), re.IGNORECASE)]
        return "\n".join(lines)
    
    output_file = os.path.splitext(input_pdf_path)[0] + ".json"
    cleaned_text = ''
    pdfReader = PdfReader(input_pdf_path)

    for i in range(len(pdfReader.pages)):
        pageObj = pdfReader.pages[i]
        raw_text = pageObj.extract_text()
        
        if raw_text:
            processed_text = clean_page_text(raw_text)
            cleaned_text += processed_text + ' '  # separate pages with newlines
    
    sentences = re.split(r'(?<=[.!?])\s+', cleaned_text.strip())
    final_sentences = [re.sub(r'\n', ' ', sentence) for sentence in sentences] # places one sentence in a line
    final_sentences = [re.sub(r'\s+', ' ', sentence).strip() for sentence in sentences] # remove extra unnecessary white spaces

    with open(output_file, "w", encoding = "utf-8") as f:
        json.dump(final_sentences, f, indent = 4, ensure_ascii = False)
    
    print(f"Sentences saved to {output_file}")

In [21]:
pdf_folder_path = "datasets/"
for filename in os.listdir(pdf_folder_path):
    if filename.lower().endswith('.pdf'):
        input_pdf_path = os.path.join(pdf_folder_path, filename)
        convert_ocr_pdf_to_text(input_pdf_path)

### Retrieve raw_data

In [22]:
dfs = {}
pdf_folder_path = "datasets/"

for folder_name in os.listdir(pdf_folder_path):
    folder_path = os.path.join(pdf_folder_path, folder_name)  
    if os.path.isdir(folder_path):  
        json_files = [f for f in os.listdir(folder_path) if f.endswith('.json')]
        for file in json_files:
            file_path = os.path.join(folder_path, file)
            print(f"Processing file: {file_path}")
            
            file_path = os.path.join(folder_path, file)
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)  # Load JSON file
            if isinstance(data, list):  
                df = pd.DataFrame(data, columns=["esg_text"])
            else:
                print(f"Skipping {file}: Unsupported format")
                continue
            
            dfs[file] = df

In [23]:
len(dfs)

0

In [24]:
for file, df in dfs.items():
    match = re.match(r"(.+?)_(\d{4})_ocr.json", file)  
    if match:
        company_name, year = match.groups()
        df["Company"] = company_name
        df["Year"] = int(year)

# Concatenate all DataFrames into df_combined
df_combined = pd.concat(dfs.values(), ignore_index=True)
missing_rows = df_combined[df_combined.isnull().any(axis=1)]
print(len(missing_rows))
df_combined.drop_duplicates(inplace=True)

ValueError: No objects to concatenate

In [28]:
df_combined["esg_text"].apply(type).value_counts()

esg_text
<class 'str'>    1193
Name: count, dtype: int64

# RAG Pipeline Architecture 

## Data Loading 
Data loading methodology: Sentence splitting, saved into each row of the csv 
Use ChromaDB for vector database: Generate embeddings for each sentence, store embeddings in vector DB, and generate embedding for the query to retrieve top-k similar chunks 

### Vector database
When you store documents in ChromaDB using collection.add(), it:

1. Generates vector embeddings for your text (if you haven't provided them).
2. Stores the document along with its embedding in the vector database.
3. Matches queries based on similarity search (cosine similarity by default).


In [ ]:
df = pd.read_csv("combined_pdfs.csv")

OSError: [Errno 22] Invalid argument: '..\x0ciles\\labeled_pdfs_1003'

In [ ]:
client = chromadb.PersistentClient(path="./chroma_db")  # Stores DB in ./chroma_db
collection = client.get_or_create_collection(name="dsa4265_ass2")
logging.basicConfig(level=logging.WARNING)

for index, row in tqdm(df.iterrows(), total=len(df), desc="Adding documents", unit="document", leave=True, ncols=100):
    doc_text = row["esg_text"]  
    doc_company = row["Company"]  
    doc_year = row["Year"]  
    doc_id = f"doc_{index}"  

    collection.add(
        ids=[doc_id], 
        documents=[doc_text],  
        metadatas=[{"company": doc_company, "year": doc_year}] 
    )

In [ ]:
print(collection.count())  

## Retrieval
Method: Semantic similarity to compare embeddings of the query to the sentences, and retrieve the top sentences with the highest similarity scores.
Limitations: Different words may have different meanings under different contexts. 

### Retrieval & Query Handling

Hybrid Search pipeline

Step 1: Vector Search in ChromaDB → Retrieve top-k relevant documents based on semantic similarity.

Step 2: BGE Reranker Search -> Evaluates how relevant the retrieved documents are by comparing each document with the query, scoring them, and reordering them according to relevant. 

In [31]:
client = chromadb.PersistentClient(path="./chroma_db")
collection = client.get_or_create_collection(name="dsa4265_ass2")

In [33]:
query = "What is the carbon emission goals of pfizer"
results = collection.query(
    query_texts=[query],
    where={"company": "Pfizer"},
    n_results=5
)

In [34]:
results

{'ids': [['doc_10665', 'doc_10666', 'doc_9646', 'doc_9655', 'doc_9648']],
 'embeddings': None,
 'documents': [['As part of the  commitment, Pfizer aims to decrease our GHG emissions by  95% and value chain emissions by 90% from 2019 levels by  2040 through accelerating the transition away from fossil  fuels and engaging suppliers to catalyze equivalent action.',
   'In 2020, after achieving our third generation GHG emission  reduction goal, Pfizer set a new goal to reduce Scope 1 & 2  GHG emissions 46% by 2030 from a 2019 baseline and, by  2025, advancing goals to reduce emissions in three Scope  3 categories.',
   'Pfizer is continuing its near-term commitment to reduce  company Greenhouse Gas (GHG) emissions aligned witha  1.5°C trajectory and to engage suppliers so that they also  set science-based GHG emissions reduction goals.',
   'Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) greenhouse gas (GHG) emissions and a 90% reduction in value chain  (Scope 3) emissions

### Reranking
Rerank documents using bge-reranker

In [35]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
retrieved_docs = [doc for doc in results["documents"][0]]
model_name = "BAAI/bge-reranker-base"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)


inputs = tokenizer(
    [query] + retrieved_docs,  
    padding=True, truncation=True, return_tensors="pt"
)

# Compute relevance scores
with torch.no_grad():
    scores = model(**inputs).logits.squeeze().tolist()

# Sort docs by relevance score
reranked_docs = [doc for _, doc in sorted(zip(scores[1:], retrieved_docs), reverse=True)]
print("Reranked Docs:", reranked_docs)

Reranked Docs: ['Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) greenhouse gas (GHG) emissions and a 90% reduction in value chain  (Scope 3) emissions by 2040’.', 'In 2020, after achieving our third generation GHG emission  reduction goal, Pfizer set a new goal to reduce Scope 1 & 2  GHG emissions 46% by 2030 from a 2019 baseline and, by  2025, advancing goals to reduce emissions in three Scope  3 categories.', 'Pfizer is continuing its near-term commitment to reduce  company Greenhouse Gas (GHG) emissions aligned witha  1.5°C trajectory and to engage suppliers so that they also  set science-based GHG emissions reduction goals.', 'As part of the  commitment, Pfizer aims to decrease our GHG emissions by  95% and value chain emissions by 90% from 2019 levels by  2040 through accelerating the transition away from fossil  fuels and engaging suppliers to catalyze equivalent action.', 'By 2040 Pfizer  aims to decrease its company GHG emissions by 95% and  its value chain emi

## Generator
Use DeepSeek API 

In [36]:
client = OpenAI(
    base_url="https://openrouter.ai/api/v1",
    api_key=API_KEY
)

In [37]:
def generate_response(query, reranked_docs):
    """Retrieve context from ChromaDB and generate an answer using DeepSeek."""
    
    context = "\n\n".join(reranked_docs)

    prompt = f"""You are an expert in ESG analysis. Please reason through step by step and then provide the final answer to the query. 
    Please verify your answer against the context provided, and rewrite the answer if inconsistent. Below is a question and relevant retrieved documents.
    
    Question: {query}

    Context:
    {context}

    Please provide a factually accurate response. If a fact is used from a document, include '(ChunkID)' next to it.
    """

    retries = 3
    while retries > 0:
        try:
            completion = client.chat.completions.create(
                model="deepseek/deepseek-r1-zero:free",
                messages=[
                    {"role": "system", "content": "You are an expert in ESG analysis. Answer factually and ensure consistency with the provided context, especially focusing on environemntal, sustainability and governance principles."},
                    {"role": "user", "content": prompt}
                ]
            )

            if completion and completion.choices and completion.choices[0].message:
                return completion.choices[0].message.content  # Return model's response

            print("Warning: Empty response. Retrying...")
            retries -= 1
            time.sleep(5)

        except Exception as e:
            print(f"Error: {e}. Retrying...")
            retries -= 1
            time.sleep(5)

    return "Error: Unable to generate a response."

In [38]:
response = generate_response(query, reranked_docs)
print("\n🔹 Generated Answer:\n", response)


🔹 Generated Answer:
 \boxed{Pfizer aims to reduce its company (Scope 1 & 2) GHG emissions by 46% by 2030 from a 2019 baseline. By 2040, Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) GHG emissions and a 90% reduction in value chain (Scope 3) emissions from 2019 levels.}


### Generating a response from Deepseek without the retrieval step's context

In [39]:
def generate_response_without_context(query):
    """Retrieve context from ChromaDB and generate an answer using DeepSeek."""

    prompt = f"""You are an expert in ESG analysis. Please reason through step by step and then provide the final answer to the query. 
    Please verify your answer against the context provided, and rewrite the answer if inconsistent. Below is a question and relevant retrieved documents.
    
    Question: {query}

    Please provide a factually accurate response. 
    """

    retries = 3
    while retries > 0:
        try:
            completion = client.chat.completions.create(
                model="deepseek/deepseek-r1-zero:free", #deepseek-reasoner #deepseek/deepseek-r1:free
                messages=[
                    {"role": "system", "content": "You are an expert in ESG analysis. Answer factually and ensure consistency with the provided context, especially focusing on environemntal, sustainability and governance principles."},
                    {"role": "user", "content": prompt}
                ]
            )

            if completion and completion.choices and completion.choices[0].message:
                return completion.choices[0].message.content  

            print("Warning: Empty response. Retrying...")
            retries -= 1
            time.sleep(5)

        except Exception as e:
            print(f"Error: {e}. Retrying...")
            retries -= 1
            time.sleep(5)

    return "Error: Unable to generate a response."

In [40]:
generate_response_without_context(query)

'\\boxed{\n"Pfizer aims to reduce Scope 1 GHG emissions by 46% by 2030 from a 2019 baseline and Scope 2 GHG emissions by 20% by the same year from the same baseline. Additionally, the company aims for a 90% absolute reduction in operational Scope 1 and Scope 2 GHG emissions combined by 2040 from a 2019 baseline."\n}'

# Post processing

Checking for hallucination, irrelevance, bias 
In this assignment, I felt that biasness wasn't really a metric required, I think it would be good to add biasness if i extracted data from third party sources grading the company esg scores. I can then compare the third-party metrics and scoring to each company's esg reports, and check if there is biasness in terms of their ratings, towards a particular, company or industry, etc. Therefore, I just added the metric for future reference, but it is not required in this assignment.

### Hallucination detection (Faithfullness)

In [3]:
def normalize_text(text):
    """Normalize text by converting to lowercase and removing punctuation."""
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)  # Remove punctuation
    return text

def fuzzy_match(sentence, doc, threshold=80):
    """Check if sentence has a fuzzy match in the document."""
    return fuzz.partial_ratio(normalize_text(sentence), normalize_text(doc)) >= threshold

def verify_facts(response, reranked_docs, fuzzy_threshold=80):
    """Detect hallucinations by checking if sentences exist in retrieved docs using fuzzy matching."""
    missing_facts = []
    
    # Split response into sentences and check if they appear in any of the documents
    for sent in response.split(". "):
        found = any(fuzzy_match(sent, doc, fuzzy_threshold) for doc in reranked_docs)
        if not found:
            missing_facts.append(sent)

    if missing_facts:
        print("Warning: Some statements are not found in the retrieved context:")
        for fact in missing_facts:
            print(f"- {fact}")
    
    return 1 - len(missing_facts) / len(response.split(". "))  # Faithfulness Score

faithfulness_score = verify_facts(response, reranked_documents)
print(f"Faithfulness Score: {faithfulness_score}")

Faithfulness Score: 1.0


## Irrelevance Check

In [11]:
model = SentenceTransformer('all-MiniLM-L6-v2')

def check_relevance(query, response, threshold=0.6):
    """
    Check the relevance of the response to the query using semantic similarity.
    """
    query_embedding = model.encode([query])
    response_embedding = model.encode([response])

    similarity = cosine_similarity(query_embedding, response_embedding)[0][0]

    if similarity >= threshold:
        return True, similarity  
    else:
        return False, similarity  

In [12]:
is_relevant = check_relevance(query, response)
print(f"Is the response relevant? {is_relevant}")

Is the response relevant? (True, 0.6938479)


In [13]:
model_name = "BAAI/bge-reranker-base" 
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

def check_bias(text):
    """
    Check for potential bias in the text using a pretrained model.
    """
    inputs = tokenizer(text, return_tensors="pt", padding=True, truncation=True)

    with torch.no_grad():
        logits = model(**inputs).logits

    # Assuming binary classification (0 = no bias, 1 = biased)
    predicted_class = torch.argmax(logits, dim=1).item()
    
    return predicted_class == 1  # 1 indicates bias (this depends on the model's labeling)

# Example Usage
response = "Pfizer has been focusing on improving diversity in their clinical trials and sharing their insights with others as part of their diversity and inclusion initiatives in 2022."

is_biased = check_bias(response)
print(f"Is the response biased? {is_biased}")

Is the response biased? False


## Retriever Evaluation
Typical metrics: RecalL@k, Precision @k, Mean Reciprocal Rank, Mean Average Precision

### Recall@k

In [ ]:
def recall_at_k(reranked_docs, ground_truth_docs, k=5):
    """Compute Recall@K: how many relevant docs were retrieved."""
    relevant_retrieved = sum(1 for doc in reranked_docs[:k] if doc in ground_truth_docs)
    return relevant_retrieved / len(ground_truth_docs) if ground_truth_docs else 0

recall_at_k(reranked_docs, ground_truth_docs, k=5) ## Need to manually label the ground truth docs. 

### MRR

In [ ]:
def mean_reciprocal_rank(reranked_docs, ground_truth_docs):
    """Compute MRR: rewards ranking of first relevant document."""
    for i, doc in enumerate(reranked_docs):
        if doc in ground_truth_docs:
            return 1 / (i + 1)  # Rank starts from 1
    return 0

### Cosine similarity

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

def compute_retrieval_relevance(query, reranked_docs):
    """Compute semantic similarity between query and retrieved docs."""
    corpus = [query] + reranked_docs
    vectorizer = TfidfVectorizer()
    tfidf_matrix = vectorizer.fit_transform(corpus)
    similarity_scores = cosine_similarity(tfidf_matrix[0], tfidf_matrix[1:])
    return similarity_scores.mean()  # Average similarity

In [16]:
compute_retrieval_relevance(query, reranked_documents)

0.09264488847701834

## Generator Evaluation 
Typical metrics: ROUGE, BLEU, BERTScore, domain-specific or task-specific metrics

### BLEU Score (Text similarity)

In [17]:
from nltk.translate.bleu_score import sentence_bleu

def compute_bleu_score(reference, generated_response):
    """Compare generated response against reference text using BLEU score."""
    reference_tokens = reference.lower().split()
    generated_tokens = generated_response.lower().split()
    return sentence_bleu([reference_tokens], generated_tokens)

In [18]:
compute_bleu_score(query, response)

/Users/chienshiyun/miniconda3/envs/4265_assignment1/lib/python3.9/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 2-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/Users/chienshiyun/miniconda3/envs/4265_assignment1/lib/python3.9/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/Users/chienshiyun/miniconda3/envs/4265_assignment1/lib/python3.9/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-

9.50440384721771e-232

### Retrieval score (relevance)

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

def compute_retrieval_relevance(reranked_docs, response):
    """Calculate how relevant the response is to the retrieved documents."""
    corpus = reranked_docs + [response]  # Combine all docs and response
    vectorizer = TfidfVectorizer()
    tfidf_matrix = vectorizer.fit_transform(corpus)
    similarity_matrix = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])
    return similarity_matrix.mean()  # Average similarity score


In [ ]:
compute_retrieval_relevance(reranked_docs, response)

### Judge LM 

In [21]:
client = genai.Client(api_key=google_api_key)
reranked_docs_str = "\n".join(reranked_documents)

gemini_eval = client.models.generate_content(
    model="gemini-2.0-flash",
    contents=f"""
                Evaluate how well the response answers the query, giving an explanation of how it answers the question, and whether the response is factually correct based on the context provided.
                I have added the query, response and retrieved context below.
                
                Query: 
                {query}
                
                Response:
                {response}
                
                Retrieved Context:
                {reranked_docs_str}
                
                Give a score from 0 to 10, and a detailed explanation on the score, where:
                - 10 = Perfectly accurate
                - 0 = Completely incorrect
                """
)

print(gemini_eval.text)

The response accurately and completely answers the query about Pfizer's carbon emission goals. It provides both the 2030 and 2040 targets for Scope 1 & 2 emissions, as well as the 2040 target for Scope 3 emissions, all of which are supported by the provided context.

Score: 10



# Conclusion

Difficulties faced and next steps: 
- I had not enough local storage space to download and extract the data of all company esg reports, so I only did a few samples. But ideally, the code should run automatically yearly to pull the latest data from all companies, and store them in a cloud storage database instead of in a local directory.
- Deciding on the how large the chunks should be. In this assignment, I have separarated them by each sentences. However, I realised that there may be more than one sentence that could answer a query, and that some sentences may provide context for the next. Therefore, I am not sure what would be the optimal number of sentences to chunk together. If I have more time in future, I would classify each sentence, and link sentences that have the same category label sequentially together as one single chunk, for better context for the RAG application. 
- Some of the input metadata from the pdf ocr reader could not be copied because it is not permitted in PDF/A. (You may wish to examine the output PDF's XMP metadata.)

Potential extensions: 
- Add more data sources, such as third party esg scoring websites and esg related news articles instead of just the company's esg reports, as the reports will be biased towards the good things that the company has done, and does not give a comprehensive view of the company's esg initiatives.
- Integrate real-time news retrieval via APIs (Google News, Twitter), automated webscraping of the latest articles, etc. 
- Build an interactive Streamlit UI for user-friendly queries. 

Iterations: 
- Prompt engineering: changed the prompts in the response generation prompt multiple times to get the best results. 
- Prompt engineering: changed the query as well to be more specific, so that the retrieved documents are less ambiguous
- Hyperparameter tuning: changed the number of documents retrieved from chromadb. If it was too many, eg 10 it may get more irrelevant documents. However, since I have split the pdf into individual sentences during chunking, there may be multiple sentences that answer the query. Therefore setting the number of documents as too low will also cause the documents retrieved to not be a comprehensive answer to the query. 

In the initial retriever, the document text has close semantic similarity words wise, but did not answer the question. For eg. the query was "What are Pfizer's environmental policies" And the result with the closest similarity was "Updated 2022 data will be published on Pfizer's Environmental Sustainability page." Because the result included the words 'Pfizr' and 'environmental'. 

The full list of what was included is: 
1. 'As part of the  commitment, Pfizer aims to decrease our GHG emissions by  95% and value chain emissions by 90% from 2019 levels by  2040 through accelerating the transition away from fossil  fuels and engaging suppliers to catalyze equivalent action.'
2. 'In 2020, after achieving our third generation GHG emission  reduction goal, Pfizer set a new goal to reduce Scope 1 & 2  GHG emissions 46% by 2030 from a 2019 baseline and, by  2025, advancing goals to reduce emissions in three Scope  3 categories.'
3. 'Pfizer is continuing its near-term commitment to reduce  company Greenhouse Gas (GHG) emissions aligned witha  1.5°C trajectory and to engage suppliers so that they also  set science-based GHG emissions reduction goals.'
4. 'Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) greenhouse gas (GHG) emissions and a 90% reduction in value chain  (Scope 3) emissions by 2040’.'
5. 'By 2040 Pfizer  aims to decrease its company GHG emissions by 95% and  its value chain emissions by 90% from 2019 levels through  accelerating the transition away from fossil fuels and  engaging suppliers to catalyze equivalent action.'

After being reranked, the reranked documents are 

1. 'Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) greenhouse gas (GHG) emissions and a 90% reduction in value chain  (Scope 3) emissions by 2040’.'
2. 'In 2020, after achieving our third generation GHG emission  reduction goal, Pfizer set a new goal to reduce Scope 1 & 2  GHG emissions 46% by 2030 from a 2019 baseline and, by  2025, advancing goals to reduce emissions in three Scope  3 categories.'
3. 'Pfizer is continuing its near-term commitment to reduce  company Greenhouse Gas (GHG) emissions aligned witha  1.5°C trajectory and to engage suppliers so that they also  set science-based GHG emissions reduction goals.'
4. 'As part of the  commitment, Pfizer aims to decrease our GHG emissions by  95% and value chain emissions by 90% from 2019 levels by  2040 through accelerating the transition away from fossil  fuels and engaging suppliers to catalyze equivalent action.'
5. 'By 2040 Pfizer  aims to decrease its company GHG emissions by 95% and  its value chain emissions by 90% from 2019 levels through  accelerating the transition away from fossil fuels and  engaging suppliers to catalyze equivalent action.'


Next, we use the LLM To generate a final answer, and the generated answer answered the query much better with the document's context. It answered the question well, using the top 2 reranked documents: 
Pfizer aims to reduce its company (Scope 1 & 2) GHG emissions by 46% by 2030 from a 2019 baseline. By 2040, Pfizer aims to achieve a 95% reduction in company (Scope 1 & 2) GHG emissions and a 90% reduction in value chain (Scope 3) emissions from 2019 levels.


This is in contrast to when no context was provided, it provided data that was not accurate. That date points "reducing scope 2 GHG emissions by 20%" and "90% reduction in operational scope 1 & 2 GHG emissions" are different from what was reported in the esg report. There was no mention of Scope 2 GHG emissions being reduced by 20%, and the reduction in scope 1&2 GHG emissions by 2040 is 95% instead of 90%). 
Here is the answer of the llm without providing context: 
Pfizer aims to reduce Scope 1 GHG emissions by 46% by 2030 from a 2019 baseline and Scope 2 GHG emissions by 20% by the same year from the same baseline. Additionally, the company aims for a 90% absolute reduction in operational Scope 1 and Scope 2 GHG emissions combined by 2040 from a 2019 baseline."

Therefore, adding context from the vectordb helps to reduce hallucination.